src/librustc_mir/transform/check_consts/qualifs.rs

```diff
@@ -101,8 +94,14 @@ pub trait Qualif {
             }
 
             Operand::Constant(ref constant) => {
-                if let Some(static_) = constant.check_static_ptr(cx.tcx) {
-                    Self::in_static(cx, static_)
+                if constant.check_static_ptr(cx.tcx).is_some() {
+                    // `mir_const_qualif` does return the qualifs in the final value of a `static`,
+                    // so we could use value-based qualification here, but we shouldn't do this
+                    // without a good reason.
+                    //
+                    // Note: this uses `constant.literal.ty` which is a reference or pointer to the
+                    // type of the actual `static` item.
+                    Self::in_any_value_of_ty(cx, constant.literal.ty)
```

Comment:
Oops, this is what the `false` lower down should be too!
Also, since this wouldn't be `ConstKind::Unevaluated`, you can just have that `else` case handle both references/pointers to statics, and everything else.

Reply:
@eddyb What's the impact of this? Will you open a follow-up PR?

Reply:
I would prefer if @matthewjasper, @oli-obk or @ecstatic-morse handled it.

But overall, I believe the code today is correct in all cases on stable (and maybe even including `const` generics), it would just be good to make it more foolproof/futureproof.

Reply:
This has been addressed on master: https://github.com/rust-lang/rust/blob/36d13cb01ba6a0a9b7c13ca2b9461a111cb3e395/src/librustc_mir/transform/check_consts/qualifs.rs#L264